cpp/node_similarity_module/algorithms/node_similarity.hpp

```diff
@@ -0,0 +1,194 @@
+    }
+    const auto &prop1 = prop1_it.ValueList();
+    const auto &prop2 = prop2_it.ValueList();
+    int size1 = prop1.Size(), size2 = prop2.Size();
+    if (size1 != size2) {
+        throw mgp::ValueException("Vectors should be of the same size.");
+    }
+    return CosineFunc(prop1, prop2);
+}
+
+
+
+/*
+Extract node neighbours.
+*/
+std::set<uint64_t> GetNeighbors(std::unordered_map<uint64_t, std::set<uint64_t>> &all_node_neighbors, const mgp::Node &node) {
+    uint64_t node_id = node.Id().AsUint();
+    const auto &result_it = all_node_neighbors.find(node_id);
+    if (result_it != all_node_neighbors.end()) {
+        return result_it->second;
+    }
+    all_node_neighbors[node_id] = std::set<uint64_t>();
+    auto &ns = all_node_neighbors[node_id];
+    const auto &rels = node.OutRelationships();
+    for (const auto &rel: rels) {
+        ns.insert(rel.To().Id().AsUint());
+    }
+    return ns;
+}
+
+
+/*
+Calculates similiraty between pairs of nodes given by src_nodes and dst_nodes.
+*/
+std::vector<std::tuple<mgp::Node, mgp::Node, double>> CalculateSimilarityPairwise(const mgp::List &src_nodes, const mgp::List &dst_nodes, node_similarity_util::Similarity similarity_mode, const std::string &property = "") {
+    if (src_nodes.Size() != dst_nodes.Size()) {
+        throw mgp::ValueException("Arguments are of different size.");
+    }
+    int num_nodes = src_nodes.Size();
+    std::vector<std::tuple<mgp::Node, mgp::Node, double>> results;
+    std::unordered_map<uint64_t, std::set<uint64_t>> all_node_neighbors;
+    for (int i = 0; i < num_nodes; ++i) {
+        const mgp::Node &src_node = src_nodes[i].ValueNode(), &dst_node = dst_nodes[i].ValueNode();
+        double similarity = 0.0;
+        if (similarity_mode == node_similarity_util::Similarity::cosine) {
+            similarity = node_similarity_algs::CosineFuncWrapper(src_node, dst_node, property);
+            results.emplace_back(src_node, dst_node, similarity);   
+            continue;
+        } 
+        const auto &ns1 = GetNeighbors(all_node_neighbors, src_node);
+        const auto &ns2 = GetNeighbors(all_node_neighbors, dst_node);
+            switch (similarity_mode) {
+                case node_similarity_util::Similarity::jaccard:
+                    similarity = node_similarity_algs::JaccardFunc(ns1, ns2);
+                    break;
+                case node_similarity_util::Similarity::overlap:
+                    similarity = node_similarity_algs::OverlapFunc(ns1, ns2);
+                    break;
+                default:
+                    break; 
+            }
+        results.emplace_back(src_node, dst_node, similarity);   
+    }
+    return results;
+}
+
+
+/*
+Calculates similarity between all pairs of nodes, in a cartesian mode.
+*/
+std::vector<std::tuple<mgp::Node, mgp::Node, double>> CalculateSimilarityCartesian(const mgp::Graph &graph, node_similarity_util::Similarity similarity_mode, const std::string &property = "") {
```

Comment:
depends on how much this function is executed, does it have any performance issues?

Reply:
@antoniofilipovic there shouldn't be any bug: by using std::string we will just create a not neccessary copy of the string when the string is actually provided from a caller. You can run this code to check it:
```cpp
#include <iostream>
#include <string>

int main()
{
    const std::string &property = "232";
    std::cout << property << std::endl;
}
```

Reply:
@Josipmrden this function is only called once

Reply:
I think this is good. It's not python and the reason why it needs something is because reference type always needs a value. It is a mechanism to defend itself against null-pointers and dereferencing of them. So I'd say this is good.